File: src/main/java/io/github/svegon/capi/event/interact/StopUsingItemListener.java
package io.github.svegon.capi.event.interact;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.network.ClientPlayerInteractionManager;
import net.minecraft.entity.player.PlayerEntity;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@FunctionalInterface
public interface StopUsingItemListener {
    Event<StopUsingItemListener> EVENT = EventFactory.createArrayBacked(StopUsingItemListener.class,
            (interactionManager, player, ci) -> {}, listeners -> (interactionManager, player, ci) -> {
        for (StopUsingItemListener listener : listeners) {
            listener.onStoppingUsingItem(interactionManager, player, ci);

            if (ci.isCancellable()) {
                return;
            }
        }
    });

    void onStoppingUsingItem(ClientPlayerInteractionManager interactionManager, PlayerEntity player, CallbackInfo ci);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/interact/StopUsingItemListener.kt
package io.github.svegon.capi.event.interact

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.network.ClientPlayerInteractionManager

fun interface StopUsingItemListener {
    fun onStoppingUsingItem(
        interactionManager: ClientPlayerInteractionManager?,
        player: PlayerEntity?,
        ci: CallbackInfo?
    )

    companion object {
        @JvmField
        val EVENT: Event<StopUsingItemListener> = EventFactory.createArrayBacked<StopUsingItemListener>(
            StopUsingItemListener::class.java,
            StopUsingItemListener { interactionManager: ClientPlayerInteractionManager?, player: PlayerEntity?, ci: CallbackInfo? -> }) { listeners: Array<StopUsingItemListener> ->
            StopUsingItemListener { interactionManager: ClientPlayerInteractionManager?, player: PlayerEntity?, ci: CallbackInfo ->
                for (listener in listeners) {
                    listener.onStoppingUsingItem(interactionManager, player, ci)

                    if (ci.isCancellable()) {
                        return@StopUsingItemListener
                    }
                }
            }
        }
    }
}
